Detectors/EMCAL/reconstruction/src/CaloRawFitterGamma2.cxx

```diff
@@ -0,0 +1,172 @@
+
+    if (nsamples > 2 && maxADC < constants::OVERFLOWCUT) {
+      doParabolaFit(timeEstimate - 1, amp, time);
+      mNiter = 0;
+      fitDone = doFit_1peak(first, nsamples, amp, time, chi2);
+
+      if (!fitDone) {
+        amp = ampEstimate;
+        time = timeEstimate;
+        chi2 = 1.e9;
+      }
+
+      time += timebinOffset;
+      timeEstimate += timebinOffset;
+      ndf = nsamples - 2;
+    }
+  }
+
+  if (fitDone) {
+    float ampAsymm = (amp - ampEstimate) / (amp + ampEstimate);
+    float timeDiff = time - timeEstimate;
+
+    if ((TMath::Abs(ampAsymm) > 0.1) || (TMath::Abs(timeDiff) > 2)) {
+      amp = ampEstimate;
+      time = timeEstimate;
+      fitDone = false;
+    }
+  }
+  if (amp >= mAmpCut) {
+    if (!fitDone) {
+      std::default_random_engine generator;
+      std::uniform_real_distribution<float> distribution(0.0, 1.0);
+      amp += (0.5 - distribution(generator));
+    }
+    time = time * constants::EMCAL_TIMESAMPLE;
+    time -= mL1Phase;
+
+    return CaloFitResults(maxADC, pedEstimate, mAlgo, amp, time, (int)time, chi2, ndf);
+  }
+  return CaloFitResults(-1, -1);
+}
+
+bool CaloRawFitterGamma2::doFit_1peak(int ifirst, int nS, float& A0, float& t0, float& chi2)
+{
+  // fit using gamma-2 function 	(ORDER =2 assumed)
+
+  if (nS < 3)
+    return false;
+  if (mNiter > mNiterationsMax)
+    return false;
+
+  double expi, ti, g_i, g_1i, g_2i, g1_1, g2_i, gp_i, g2p_i, q1_i, q2_i, D, dA, dt, delta;
+  double c11 = 0;
+  double c12 = 0;
+  double c21 = 0;
+  double c22 = 0;
+  double d1 = 0;
+  double d2 = 0;
+  chi2 = 0;
+
+  mNiter++;
+
+  for (int i = 0; i < nS; i++) {
+
+    ti = (i - t0) / constants::TAU;
+    if ((ti + 1) < 0)
+      continue;
+
+    expi = TMath::Exp(-2 * ti);
+    g_2i = expi;
+    g_1i = (ti + 1) * g_2i;
+    g_i = (ti + 1) * g_1i;
+    gp_i = 2 * (g_i - g_1i);
+    g2p_i = 2 * g_i * gp_i;
+    q1_i = g_2i * (2 * ti + 1);
+    q2_i = g_1i * g_1i * (4 * ti + 1);
+    c11 += (getReversed(i) - A0 * 2 * g_i) * gp_i;
+    c12 += g_i * g_i;
+    c21 += getReversed(i) * q1_i - A0 * q2_i;
+    c22 += g_i * g_1i;
+    delta = A0 * g_i - getReversed(i);
+    d1 += delta * g_i;
+    d2 += delta * g_1i;
+    chi2 += (delta * delta);
+  }
+
+  D = c11 * c22 - c12 * c21;
+
+  if (TMath::Abs(D) < 1.e-9)
+    return false;
+
+  dt = (d1 * c22 - d2 * c12) / D * constants::TAU;
+  dA = (d1 * c21 - d2 * c11) / D;
+
+  t0 += dt;
+  A0 += dA;
+
+  bool res = true;
+
+  if (TMath::Abs(dA) > 1 || TMath::Abs(dt) > 0.01)
+    res = doFit_1peak(ifirst, nS, A0, t0, chi2);
+
+  return res;
+}
+
+void CaloRawFitterGamma2::doParabolaFit(int x, float& amp, float& time)
+{
+  // estimate initial values
+  double a = (getReversed(x + 2) + getReversed(x) - 2. * getReversed(x + 1)) / 2.;
+
+  if (TMath::Abs(a) < 1.e09) {
+    amp = getReversed(x + 1);
+    time = x + 1;
+    return;
+  }
+
+  double b = getReversed(x + 1) - getReversed(x) - a * (2. * x + 1);
+  double c = getReversed(x) - b * x - a * x * x;
```

Comment:
Parameter names should be meaningful.

Reply:
This is a parabola equation, so the parameters are called "a", "b", and "c" by convention.

Reply:
OK, as in this case the names are pretty standard we can probably live with them even though they are single letter variable names.